src/panoptes/pocs/state/machine.py

```diff
@@ -109,25 +109,41 @@ def run(self, exit_when_done=False, run_once=False, initial_next_state='ready'):
 
         self.logger.debug(f'Starting run loop')
         while self.keep_running:
+
             # BEFORE TRANSITION TO STATE
-            self.logger.info(f'Run loop: self.state={self.state!r} self.next_state={self.next_state!r}')
-
-            # Before moving to next state, check for required horizon level and wait if necessary.
-            required_horizon = self._horizon_lookup.get(self.next_state, 'observe')
-            self.logger.debug(f'Checking for required_horizon={required_horizon!r} for self.next_state={self.next_state!r}')
-            while not self.is_safe(no_warning=True, horizon=required_horizon):
-                self.logger.info(f'Waiting for required_horizon={required_horizon!r} for self.next_state={self.next_state!r}')
-                check_delay = self.get_config('wait_delay',
-                                              default=60 * 3)  # Check every 3 minutes.
+            self.logger.info(f'Run loop: self.state={self.state!r}'
+                             f'self.next_state={self.next_state!r}')
+
+            # Before moving to next state, wait for required horizon if necessary
+            while True:
+                # If not safe, go to park
+                self.is_safe(park_if_not_safe=True, ignore=['is_dark'])
+
+                # The state may have changed since the start of the while loop
+                # e.g. if self.park is called from self.is_safe
+                # So we need to check if the new state is always safe
+                if self.get_state(self.next_state).is_always_safe:
+                    break
+
+                # Check the horizon here because next state may have changed in loop
+                required_horizon = self._horizon_lookup.get(self.next_state, 'observe')
+                if self.is_dark(horizon=required_horizon):
+                    break
+                self.logger.info(f"Waiting for required_horizon={required_horizon!r} for "
+                                 f"self.next_state={self.next_state!r}")
+
+                # Sleep before checking again
+                check_delay = self.get_config('wait_delay', default=120)
```

Comment:
This does a lookup in the config server so unless we want to dynamically change it this should be moved outside the while loop ideally. Not critical.